PRODDOCS-650: Databricks OAuth flow (#9556)
fishtown-analytics/docs.getdbt.com: website/docs/docs/platform/manage-access/set-up-databricks-oauth.md

---
title: "Set up Databricks OAuth"

> description: "Learn how dbt administrators and developers can use Databricks OAuth to control access in a dbt account."

Reviewer: Love where your heads at including the end-user here, but maybe we should use the word "user" rather than "developer" to allow for some growth and expansion down the road

id: "set-up-databricks-oauth"
---

# Set up Databricks OAuth <Lifecycle status="managed, managed_plus" />

> The <Constant name="dbt_platform" /> supports developer OAuth ([OAuth for partner solutions](https://docs.databricks.com/en/integrations/manage-oauth.html)) with Databricks on Enterprise and Enterprise+ plans. When you enable Databricks OAuth for a <Constant name="dbt" /> project, all developers must authenticate with Databricks in order to use the <Constant name="studio_ide" />.

Reviewer: We should clarify that it's dbt that needs Enterprise or Enterprise+ plan and not Databricks
```suggestion
<Constant name="dbt_platform" /> accounts on Enterprise and Enterprise+ plans support developer OAuth ([OAuth for partner solutions](https://docs.databricks.com/en/integrations/manage-oauth.html)) with Databricks. When you enable Databricks OAuth for a <Constant name="dbt" /> project, all developers must authenticate with Databricks in order to use the <Constant name="studio_ide" />.
```

### Development and deployment authentication

Databricks OAuth uses user-to-machine (U2M) authentication for interactive development in the <Constant name="studio_ide" />. It applies to development credentials only.

Deployment environments use separate credentials configured at the environment level. They do not use developer OAuth. For deployment and batch jobs, configure token-based credentials in a [deployment connection profile](/docs/platform/about-profiles). A [service principal](https://docs.databricks.com/en/admin/users-groups/service-principals.html) token is recommended for production and non-production jobs.

For setup steps, refer to [Defining your dbt deployment environment](/guides/set-up-your-databricks-dbt-project#defining-your-dbt-deployment-environment). For more information on deployment environments, refer to [Deploy environments](/docs/deploy/deploy-environments).

### Configure Databricks OAuth (Databricks admin)

To get started, add dbt as a custom OAuth application in Databricks.

> The following steps use the Databricks UI. Alternatively, use the [Databricks CLI](https://docs.databricks.com/en/integrations/enable-disable-oauth.html#enable-custom-oauth-applications-using-the-cli). Use the redirect URLs in [Redirect URLs by region](#redirect-urls-by-region) for either method.

Reviewer: ```suggestion
The following steps use the Databricks UI. Alternatively, use the [Databricks CLI](https://docs.databricks.com/en/integrations/enable-disable-oauth.html#enable-custom-oauth-applications-using-the-cli). Use the [redirect URLs for your region](#redirect-urls-by-region) for either method.
```

1. Log in to the [account console](https://accounts.cloud.databricks.com/?_ga=2.255771976.118201544.1712797799-1002575874.1704693634) and click the **Settings** icon in the sidebar.

2. On the **App connections** tab, click **Add connection**.

3. Enter the following details:
   - A name for your connection.

>    - The redirect URLs for your OAuth connection. Use both URLs listed for your account region in [Redirect URLs by region](#redirect-urls-by-region).

Reviewer: We might be overlinking to the table a bit here since it's linked above and it's on the same page immediately following this. I'd either link here or link above but not both
```suggestion
   - The redirect URLs for your OAuth connection. Use both URLs listed for your account region.
```

3. Choose your project from the list.
4. Click **Connections** and select the Databricks connection.
5. Click **Edit**.
6. Under the **Optional settings** section, add the **OAuth Client ID** and **OAuth Client Secret** from the Databricks OAuth app.

<Lightbox
  src="/img/docs/dbt-platform/using-dbt-platform/dbt-platform-enterprise/DBX-auth/dbt-databricks-oauth.png"
  title="Add Databricks OAuth credentials to the connection"

>   alt="The Optional settings section of a Databricks connection in dbt, with OAuth Client ID and OAuth Client Secret fields."

Reviewer: See my comment on the next one regarding alt text

/>

### Authenticating to Databricks (Studio IDE developer)

Once the Databricks OAuth connection is set up for a <Constant name="dbt" /> project, each <Constant name="dbt" /> user will need to authenticate with Databricks in order to use the <Constant name="studio_ide" />. To do so:

1. From <Constant name="dbt" />, click on your account name in the left side menu and select **Account settings**.
2. Under **Your profile**, select **Credentials**.
3. Choose your project from the list and click **Edit**.
4. Select **OAuth** as the authentication method, and click **Save**.
5. Finalize by clicking the **Connect Databricks Account** button.

<Lightbox
  src="/img/docs/dbt-platform/using-dbt-platform/dbt-platform-enterprise/DBX-auth/dbt-databricks-oauth-user.png"
  title="Connect your Databricks account from development credentials"

>   alt="Development credentials in dbt with Auth method set to OAuth and the Connect Databricks Account button."

Reviewer: Alt text is a good idea in the long run, but we haven't historically implemented it at this point (except in some blog posts). I think we should formulate a plan to implement site-wide, but exclude it for now so we're consistent

/>

You will then be redirected to Databricks and asked to approve the connection. This redirects you back to <Constant name="dbt" />. You should now be an authenticated Databricks user, ready to use the <Constant name="studio_ide" />.

### Studio OAuth flow

> Once a developer has authorized <Constant name="dbt" /> with Databricks, Databricks returns a refresh token to the <Constant name="dbt" /> application.

Reviewer: Same comment as up top, maybe "user" would be more flexible here

> At the start of each Studio session, <Constant name="dbt" /> exchanges this refresh token for an access token, which it uses to connect to Databricks and execute queries in the <Constant name="studio_ide" /> on the developer's behalf. <Constant name="dbt" /> does not fetch new access tokens during an active session. When the access token expires (default: 60 minutes), restart the <Constant name="studio_ide" /> to continue development.

Reviewer: ```suggestion
At the start of each <Constant name="studio_ide" /> session, <Constant name="dbt" /> exchanges the refresh token for an access token, which it uses to connect to Databricks and execute queries on the developer's behalf. <Constant name="dbt" /> does not fetch new access tokens during an active session. When the access token expires (default: 60 minutes), the user must restart the <Constant name="studio_ide" /> to continue development.
```
